system-addon/common/PrerenderData.jsm

```diff
@@ -1,32 +1,73 @@
-const prefConfig = {
-  // Prefs listed with "invalidates: true" will prevent the prerendered version
+class _PrerenderData {
+  constructor(options) {
+    this.initialPrefs = options.initialPrefs;
+    this.initialSections = options.initialSections;
+    this._setValidation(options.validation);
+  }
+
+  get validation() {
+    return this._validation;
+  }
+
+  set validation(value) {
+    this._setValidation(value);
+  }
+
+  get invalidatingPrefs() {
+    return this._invalidatingPrefs;
+  }
+
+    // This is needed so we can use it in the constructor
+  _setValidation(value = []) {
+    this._validation = value;
+    this._invalidatingPrefs = value.reduce((result, next) => {
+      if (typeof next === "string") {
+        result.push(next);
+        return result;
+      } else if (next && next.oneOf) {
+        return result.concat(next.oneOf);
+      }
+      throw new Error("Your validation configuration is not properly configured");
+    }, []);
+  }
+
+  arePrefsValid(getPref) {
+    for (const prefs of this.validation) {
+      // {oneOf: ["foo", "bar"]}
+      if (prefs && prefs.oneOf && !prefs.oneOf.some(name => getPref(name) === this.initialPrefs[name])) {
+        return false;
+
+      // "foo"
+      } else if (getPref(prefs) !== this.initialPrefs[prefs]) {
+        return false;
+      }
+    }
+    return true;
+  }
+}
+
+this.PrerenderData = new _PrerenderData({
+  initialPrefs: {
+    "migrationExpired": true,
+    "showTopSites": true,
+    "showSearch": true,
+    "topSitesCount": 6,
+    "feeds.section.topstories": true,
+    "feeds.section.highlights": true
+  },
+  // Prefs listed as invalidating will prevent the prerendered version
   // of AS from being used if their value is something other than what is listed
   // here. This is required because some preferences cause the page layout to be
   // too different for the prerendered version to be used. Unfortunately, this
   // will result in users who have modified some of their preferences not being
   // able to get the benefits of prerendering.
-  "migrationExpired": {value: true},
-  "showTopSites": {
-    value: true,
-    invalidates: true
-  },
-  "showSearch": {
-    value: true,
-    invalidates: true
-  },
-  "topSitesCount": {value: 6},
-  "feeds.section.topstories": {
-    value: true,
-    invalidates: true
-  }
-};
-
-this.PrerenderData = {
-  invalidatingPrefs: Object.keys(prefConfig).filter(key => prefConfig[key].invalidates),
-  initialPrefs: Object.keys(prefConfig).reduce((obj, key) => {
-    obj[key] = prefConfig[key].value;
-    return obj;
-  }, {}), // This creates an object of the form {prefName: value}
+  validation: [
+    "showTopSites",
+    "showSearch",
+    // This means if either of these are set to their default values,
+    // prerendering can be used.
+    {oneOf: ["feeds.section.topstories", "feeds.section.highlights"]}
```

Comment:
Oh, I forgot to mention that this list doesn't include `topSitesCount` and `migrationExpired` where it did before. I would assume that was intentional.